src/bootstrap/bin/rustc.rs

```diff
@@ -125,6 +124,17 @@ fn main() {
         cmd.arg(format!("-Cdebuginfo={}", debuginfo_level));
     }
 
+    if env::var_os("RUSTC_DENY_WARNINGS").is_some() &&
+       env::var_os("RUSTC_EXTERNAL_TOOL").is_none() {
+        cmd.arg("-Dwarnings");
+        cmd.arg("-Drust_2018_idioms");
+        if stage != "0" && crate_name != Some("rustc_version") && // cfg(not(bootstrap))
```

Comment:
Can you move the cfg comment ontop of the if and expand it a bit, something like "Internal lints may change, so don't enable them on the bootstrap compiler. Also, only enable them for internal crates."

I think rustc_version is just excluded because it's actually external but happens to fall into the use_internal_crates check, could it be moved into that function and a comment added saying "this is actually an external crate"?

Reply:
>Internal lints may change, so don't enable them on the bootstrap compiler

The reason is simpler - stage 0 compiler doesn't understand `-Drustc::internal` since it's too new for it and gives an error.

`rustc_version` for some reason reports `stage == "1"` when it's actually built by a stage 0 compiler that errors on `-Drustc::internal` (this happens during `x.py doc`, IIRC), it doesn't have issues with the lints themselves.
So it's also a part of the bootstrap problem and the exception and we'll be able to remove this exception during the next bootstrap compiler update.
(There are other "false-positive" `rustc_*` crates which happen to pass the internal lint checks and therefore not listed as exceptions.)

I'll add a comment.

Reply:
I'd expect the stage 0 compiler to understand -Drustc::internal now, since we just bumped beta -- should we remove that check then?

Reply:
Hm, I think I tried it and it didn't work.
Does stage 0 include https://github.com/rust-lang/rust/pull/61545 (merged yesterday)?

Reply:
Ah, no, that's ~one day too recent :)